src/linear_algebra/determinant.lean

```diff
@@ -52,6 +55,149 @@ variables {M' : Type*} [add_comm_group M'] [module R M']
+
+/-- If `R^m` and `R^n` are linearly equivalent, then `m` and `n` are also equivalent. -/
+def equiv_of_pi_lequiv_pi {R : Type*} [integral_domain R]
+  (e : (m → R) ≃ₗ[R] (n → R)) : m ≃ n :=
+basis.index_equiv (basis.of_equiv_fun e.symm) (pi.basis_fun _ _)
+
+/-- If `M` and `M'` are each other's inverse matrices, they are square matrices up to
+equivalence of types. -/
+def matrix.equiv_of_inv [decidable_eq m] [decidable_eq n]
+  {M : matrix m n A} {M' : matrix n m A}
+  (hMM' : M ⬝ M' = 1) (hM'M : M' ⬝ M = 1) :
+  m ≃ n :=
+let e : (m → A) ≃ₗ[A] (n → A) :=
+{ to_fun := matrix.to_lin' M',
+  inv_fun := M.to_lin',
+  left_inv := λ x, by rw [← matrix.to_lin'_mul_apply, hMM', matrix.to_lin'_one, id_apply],
+  right_inv := λ x, by rw [← matrix.to_lin'_mul_apply, hM'M, matrix.to_lin'_one, id_apply],
+  .. matrix.to_lin' M' } in
+equiv_of_pi_lequiv_pi e
+
+/-- If `M'` is a two-sided inverse for `M` (indexed differently), `det (M ⬝ N ⬝ M') = det N`. -/
+lemma matrix.det_conjugate
+  [decidable_eq m] [decidable_eq n]
+  {M : matrix m n A} {M' : matrix n m A} {N : matrix n n A}
+  (hMM' : M ⬝ M' = 1) (hM'M : M' ⬝ M = 1) :
+  det (M ⬝ N ⬝ M') = det N :=
+begin
+  letI := classical.dec_eq A,
+  -- Although `m` and `n` are different a priori, we will show they have the same cardinality.
+  -- This turns the problem into one for square matrices, which is easy.
+  let e : m ≃ n := matrix.equiv_of_inv hMM' hM'M,
+  have : det (matrix.reindex_linear_equiv e (equiv.refl _) M ⬝ N ⬝
+              matrix.reindex_linear_equiv (equiv.refl _) e M') =
+         det N,
+  { rw [det_mul, det_mul, mul_comm, ← mul_assoc, ← det_mul, ← matrix.reindex_linear_equiv_mul,
+        matrix.reindex_linear_equiv_refl_refl, hM'M, linear_equiv.refl_apply, det_one, one_mul] },
+  convert this,
+  rw [← matrix.det_reindex_linear_equiv_self e (M ⬝ N ⬝ M'),
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) e,
+      matrix.reindex_linear_equiv_mul e (equiv.refl n) (equiv.refl n),
+      matrix.reindex_linear_equiv_refl_refl, linear_equiv.refl_apply]
+end
+
+end conjugate
+
+namespace linear_map
+
+/-! ### Determinant of a linear map -/
+
+variables {A : Type*} [integral_domain A] [module A M]
+variables {κ : Type*} [fintype κ]
+
+/-- The determinant of `linear_map.to_matrix` does not depend on the choice of basis. -/
+lemma det_to_matrix_eq_det_to_matrix [decidable_eq κ]
+  (b : basis ι A M) (c : basis κ A M) (f : M →ₗ[A] M) :
+  det (linear_map.to_matrix b b f) = det (linear_map.to_matrix c c f) :=
+by rw [← linear_map_to_matrix_mul_basis_to_matrix c b c,
+       ← basis_to_matrix_mul_linear_map_to_matrix b c b,
+       matrix.det_conjugate]; rw [basis.to_matrix_mul_to_matrix, basis.to_matrix_self]
+
+/-- The determinant of an endomorphism given a basis.
+
+See `linear_map.det` for a version that does not depend on a choice of basis.
+-/
+protected def det_aux (b : basis ι A M) (f : M →ₗ[A] M) : A :=
+matrix.det (linear_map.to_matrix b b f)
```

Comment:
Is it worth giving this the monoid_hom structure immediately?
```suggestion
def det_aux (b : basis ι A M) : (M →ₗ[A] M) →* A :=
(monoid_hom.of (matrix.det)).comp (to_matrix_alg_equiv b : (M →ₗ[A] M) →* matrix ι ι A)
```
It breaks dot notation in a few places below, but otherwise seems to work fine. You might be able to eliminate a handful of proofs too.

Reply:
You could even fold `det_aux_eq` into the same definition:
```lean
/-- The determinant of an endomorphism given a basis.

See `linear_map.det` for a version that does not depend on a choice of basis.
-/
def det_aux : trunc (basis ι A M) → (M →ₗ[A] M) →* A :=
trunc.lift
  (λ b : basis ι A M,
    (monoid_hom.of (matrix.det)).comp (to_matrix_alg_equiv b : (M →ₗ[A] M) →* matrix ι ι A))
  (λ b c, begin
    ext f,
    dsimp [to_matrix_alg_equiv],
    rw det_to_matrix_eq_det_to_matrix b c,
  end)
```

Reply:
I didn't go for fancy types since I expect `det_aux` won't be used a lot anyway. In any case, the `monoid_hom` did help a bit with conciseness. It felt like `trunc` didn't have enough API to make the change worth it on its own, but here it is anyway :)